Add original error context info to GemRequireError message

RochesterinNYC opened the pull request:
Addresses #4182.

RochesterinNYC commented:
Hm, actually, @indirect, any tips on a quick way to reproduce the error condition without requiring something running stuff in a Docker container or uninstalling my node environment? I'd like to check the formatting and see if it's clean looking.

segiddins commented on lib/bundler/errors.rb:
```diff
@@ -41,7 +41,10 @@ class GemRequireError < BundlerError
-      super(msg)
+      full_message = msg + "\nError was: #{orig_exception.message}\n"\
+                      "Backtrace for gem load error was:\n"\
+                      "#{orig_exception.backtrace.inspect}"
```
Backtrace should be joined rather than inspected.

segiddins commented:
Build a gem that just raises in the file bundler will attempt to require?

RochesterinNYC commented:
Got y'all. Updated the PR with a change to the appropriate spec and updated the formatting.
The output will now look like:
```
/path/bundler/lib/bundler/runtime.rb:80:in `rescue in block (2 levels) in require': There was an error while trying to load the gem 'faulty'. (Bundler::GemRequireError)
Gem Load Error is: Gem Internal Error Message
Backtrace for gem load error is:
/path/bundler/tmp/libs/faulty-1.0.0/lib/faulty.rb:1:in `<top (required)>'
/path/bundler/lib/bundler/runtime.rb:77:in `require'
/path/bundler/lib/bundler/runtime.rb:77:in `block (2 levels) in require'
/path/bundler/lib/bundler/runtime.rb:72:in `each'
/path/bundler/lib/bundler/runtime.rb:72:in `block in require'
/path/bundler/lib/bundler/runtime.rb:61:in `each'
/path/bundler/lib/bundler/runtime.rb:61:in `require'
/path/bundler/lib/bundler.rb:100:in `require'
-e:2:in `<main>'
Bundler Error Backtrace:
  from /path/bundler/lib/bundler/runtime.rb:76:in `block (2 levels) in require'
  from /path/bundler/lib/bundler/runtime.rb:72:in `each'
  from /path/bundler/lib/bundler/runtime.rb:72:in `block in require'
  from /path/bundler/lib/bundler/runtime.rb:61:in `each'
  from /path/bundler/lib/bundler/runtime.rb:61:in `require'
  from /path/bundler/lib/bundler.rb:100:in `require'
  from -e:2:in `<main>'
```
The backtrace for the gem load error would, for the `coffee-rails` example that the original issue #4182 brought up, be all stuff in the internals of the coffee-rails gem presumably. It's only still bundler-related code/files in backtrace for this example output because this is copied from the test suite.

segiddins commented:
👍🏻

indirect commented:
@homu r+

homu commented:
:pushpin: Commit 85d6cea has been approved by `indirect`

<!-- @homu r=indirect 85d6ceae60cb1be0ca230c93c93bec90675e2145 -->

homu commented:
:hourglass: Testing commit 85d6cea with merge ba0bac9...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/bundler/bundler/builds/98953443)

emirkin commented:
@RochesterinNYC what is the simplest way to pick up this fix considering that the new version of Bundler won't be released for a while?

RochesterinNYC commented:
@emirkin a new release of `v1.12.0` will be made very soon.

If you really don't want to wait for it, though, you can install `v1.12.0.pre.2` via `gem install bundler --pre`

emirkin commented:
@RochesterinNYC Thanks!
